src/lib.rs

```diff
@@ -195,38 +210,67 @@ where
     Ok(result)
 }
 
+fn close(event_loop: &PyAny) -> PyResult<()> {
+    event_loop.call_method1(
+        "run_until_complete",
+        (event_loop.call_method0("shutdown_asyncgens")?,),
+    )?;
+
+    // how to do this prior to 3.9?
+    if event_loop.hasattr("shutdown_default_executor")? {
+        event_loop.call_method1(
+            "run_until_complete",
+            (event_loop.call_method0("shutdown_default_executor")?,),
+        )?;
+    }
+
+    event_loop.call_method0("close")?;
+
+    Ok(())
+}
+
 /// Attempt to initialize the Python and Rust event loops
 ///
 /// - Must be called before any other pyo3-asyncio functions.
 /// - Calling `try_init` a second time returns `Ok(())` and does nothing.
 ///   > In future versions this may return an `Err`.
+#[deprecated(since = "0.14.0")]
 pub fn try_init(py: Python) -> PyResult<()> {
-    EVENT_LOOP.get_or_try_init(|| -> PyResult<PyObject> {
-        let asyncio = py.import("asyncio")?;
-        let ensure_future = asyncio.getattr("ensure_future")?;
-        let event_loop = asyncio.call_method0("get_event_loop")?;
+    CACHED_EVENT_LOOP.get_or_try_init(|| -> PyResult<PyObject> {
+        let event_loop = asyncio_get_event_loop(py)?;
         let executor = py
             .import("concurrent.futures.thread")?
-            .getattr("ThreadPoolExecutor")?
-            .call0()?;
+            .call_method0("ThreadPoolExecutor")?;
         event_loop.call_method1("set_default_executor", (executor,))?;
-        let call_soon = event_loop.getattr("call_soon_threadsafe")?;
-        let create_future = event_loop.getattr("create_future")?;
-
-        ASYNCIO.get_or_init(|| asyncio.into());
-        ENSURE_FUTURE.get_or_init(|| ensure_future.into());
-        EXECUTOR.get_or_init(|| executor.into());
-        CALL_SOON.get_or_init(|| call_soon.into());
-        CREATE_FUTURE.get_or_init(|| create_future.into());
+
+        EXECUTOR.set(executor.into()).unwrap();
         Ok(event_loop.into())
     })?;
 
     Ok(())
 }
 
+fn asyncio(py: Python) -> PyResult<&PyAny> {
+    ASYNCIO
+        .get_or_try_init(|| Ok(py.import("asyncio")?.into()))
+        .map(|asyncio| asyncio.as_ref(py))
+}
+
+fn asyncio_get_event_loop(py: Python) -> PyResult<&PyAny> {
+    asyncio(py)?.call_method0("get_event_loop")
```

Comment:
Shouldn't we make a custom wrapper type for event loops that is `!Send`, to ensure that methods on it aren't called from alternative threads? Just an idea, to guard against footguns.

Reply:
Making the event loop wrapper `!Send` probably wouldn't allow us to signal completion on a `tokio` or `async_std` thread. This library _should_ guard against this issue by proxying everything through `call_soon_threadsafe` when it needs to complete. 

In principle I don't have a problem with getting rid of potential footguns, I'm just not sure how to go about doing it here.

Reply:
Good point, thanks for replying though